array.go

```diff
@@ -521,6 +521,36 @@ func toIdx(v valueInt) uint32 {
 	return math.MaxUint32
 }
 
+// array length threshold above which we should estimate mem usage
+var arrayLenThreshold = 1_000
+
+// for very large arrays calculating mem usage for each item becomes
+// expensive both in terms of memory used by the host to compute it
+// and timeout. With this function we grab a sample of 10% items
+// determine their average mem usage and use that to estimate mem
+// usage of the whole array
+func estimateMemUsage(ctx *MemUsageContext, values []Value) (uint64, error) {
+	var total, samplesVisited uint64
+	var averageMemUsage float32
+	sampleSize := len(values) / 10
+
+	for i := 0; i < len(values); i += sampleSize {
```

Comment:
I think we would want a random sampling, not just the first 10 to make this estimation less prone to malicious intent.

Reply:
This would grab one item every `sampleSize`, I thought about random sampling but then our memory usage would be different for every function run. I can see customers coming up asking why their mem usage is different for every call.

Reply:
True, great point.